core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -1852,14 +1947,22 @@ private long findLastOffsetAcknowledged() {
         long lastOffsetAcknowledged = -1;
         try {
             for (NavigableMap.Entry<Long, InFlightBatch> entry : cachedState.entrySet()) {
+                if (initialReadGapOffset != null && entry.getKey() >= initialReadGapOffset.gapStartOffset()) {
+                    break;
+                }
                 InFlightBatch inFlightBatch = entry.getValue();
                 if (inFlightBatch.offsetState() == null) {
                     if (!isRecordStateAcknowledged(inFlightBatch.batchState())) {
                         return lastOffsetAcknowledged;
                     }
-                    lastOffsetAcknowledged = inFlightBatch.lastOffset();
+                    if (initialReadGapOffset == null || inFlightBatch.lastOffset() < initialReadGapOffset.gapStartOffset()) {
+                        lastOffsetAcknowledged = inFlightBatch.lastOffset();
+                    }
                 } else {
                     for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+                        if (initialReadGapOffset != null && offsetState.getKey() >= initialReadGapOffset.gapStartOffset()) {
+                            break;
+                        }
```

Comment:
Can we write comments here please. Do we really require 3 checks? The only thing you need is to identify when to stop either when saw some gap or unacked data.